test/test_common/simulated_time_system.cc

```diff
@@ -215,58 +215,77 @@ SimulatedTimeSystemHelper::~SimulatedTimeSystemHelper() { --instance_count; }
 bool SimulatedTimeSystemHelper::hasInstance() { return instance_count > 0; }
 
 SystemTime SimulatedTimeSystemHelper::systemTime() {
-  Thread::LockGuard lock(mutex_);
+  absl::MutexLock lock(&mutex_);
   return system_time_;
 }
 
 MonotonicTime SimulatedTimeSystemHelper::monotonicTime() {
-  Thread::LockGuard lock(mutex_);
+  absl::MutexLock lock(&mutex_);
   return monotonic_time_;
 }
 
-void SimulatedTimeSystemHelper::sleep(const Duration& duration) {
+void SimulatedTimeSystemHelper::advanceTimeAsync(const Duration& duration) {
   only_one_thread_.checkOneThread();
-  mutex_.lock();
+  absl::MutexLock lock(&mutex_);
   MonotonicTime monotonic_time =
       monotonic_time_ + std::chrono::duration_cast<MonotonicTime::duration>(duration);
-  setMonotonicTimeAndUnlock(monotonic_time);
+  setMonotonicTimeLockHeld(monotonic_time);
+}
+
+void SimulatedTimeSystemHelper::advanceTimeWait(const Duration& duration) {
+  only_one_thread_.checkOneThread();
+  absl::MutexLock lock(&mutex_);
+  MonotonicTime monotonic_time =
+      monotonic_time_ + std::chrono::duration_cast<MonotonicTime::duration>(duration);
+  setMonotonicTimeLockHeld(monotonic_time);
+  waitForNoPendingLockHeld();
+}
+
+void SimulatedTimeSystemHelper::waitForNoPendingLockHeld() const EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+  mutex_.Await(absl::Condition(
+      +[](const uint32_t* pending_alarms) -> bool { return *pending_alarms == 0; },
+      &pending_alarms_));
 }
 
 Thread::CondVar::WaitStatus SimulatedTimeSystemHelper::waitFor(
     Thread::MutexBasicLockable& mutex, Thread::CondVar& condvar,
     const Duration& duration) noexcept EXCLUSIVE_LOCKS_REQUIRED(mutex) {
   only_one_thread_.checkOneThread();
+
+  // TODO(#10568): This real-time polling delay should not be necessary. Without
+  // it, test/extensions/filters/http/cache:cache_filter_integration_test fails
+  // about 40% of the time.
   const Duration real_time_poll_delay(
       std::min(std::chrono::duration_cast<Duration>(std::chrono::milliseconds(50)), duration));
   const MonotonicTime end_time = monotonicTime() + duration;
 
-  while (true) {
+  bool cont = true;
+  while (cont) {
     // First check to see if the condition is already satisfied without advancing sim time.
     if (condvar.waitFor(mutex, real_time_poll_delay) == Thread::CondVar::WaitStatus::NoTimeout) {
       return Thread::CondVar::WaitStatus::NoTimeout;
     }
 
-    // Wait for the libevent poll in another thread to catch up prior to advancing time.
-    if (hasPending()) {
-      continue;
-    }
-
-    mutex_.lock();
-    if (monotonic_time_ < end_time) {
-      if (alarms_.empty()) {
-        // If no alarms are pending, sleep till the end time.
-        setMonotonicTimeAndUnlock(end_time);
+    mutex.unlock();
```

Comment:
can you comment on the unusual lock pattern?

Reply:
Added comment:
```
    // This function runs with the caller-provided mutex held. We need to
    // hold this->mutex_ while accessing the timer-queue and blocking on
    // callbacks completing. To avoid potential deadlock we must drop
    // the caller's mutex before taking ours. We also must care to avoid
    // break/continue/return/throw during this non-RAII lock operation.
```